pjmedia/src/pjmedia/txt_stream.c

```diff
@@ -0,0 +1,1054 @@
+    char *p;
+    unsigned buf_size;
+    pj_status_t status;
+    pjmedia_transport_attach_param att_param;
+    pjmedia_clock_param cparam;
+
+    PJ_ASSERT_RETURN(endpt && info && tp && p_stream, PJ_EINVAL);
+
+    if (pool == NULL) {
+        own_pool = pjmedia_endpt_create_pool( endpt, "tstrm%p",
+                                              2000, 2000);
+        PJ_ASSERT_RETURN(own_pool != NULL, PJ_ENOMEM);
+        pool = own_pool;
+    }
+
+    /* Allocate the text stream: */
+    stream = PJ_POOL_ZALLOC_T(pool, pjmedia_txt_stream);
+    PJ_ASSERT_RETURN(stream != NULL, PJ_ENOMEM);
+    c_strm = &stream->base;
+    c_strm->own_pool = own_pool;
+
+    /* Duplicate stream info */
+    pj_memcpy(&stream->si, info, sizeof(*info));
+    c_strm->si = (pjmedia_stream_info_common *)&stream->si;
+    pj_strdup(pool, &stream->si.fmt.encoding_name, &info->fmt.encoding_name);
+    pjmedia_rtcp_fb_info_dup(pool, &c_strm->si->loc_rtcp_fb,
+                             &info->loc_rtcp_fb);
+    pjmedia_rtcp_fb_info_dup(pool, &c_strm->si->rem_rtcp_fb,
+                             &info->rem_rtcp_fb);
+
+    /* Init stream/port name */
+    name.ptr = (char*) pj_pool_alloc(pool, M);
+    name.slen = pj_ansi_snprintf(name.ptr, M, "tstrm%p", stream);
+    c_strm->port.info.name = name;
+
+    /* Init stream: */
+    c_strm->endpt = endpt;
+    c_strm->dir = info->dir;
+    c_strm->user_data = user_data;
+    c_strm->rtcp_interval = (PJMEDIA_RTCP_INTERVAL-500 + (pj_rand()%1000)) *
+                            info->fmt.clock_rate / 1000;
+    c_strm->rtcp_sdes_bye_disabled = info->rtcp_sdes_bye_disabled;
+
+    c_strm->jb_last_frm = PJMEDIA_JB_NORMAL_FRAME;
+    c_strm->rtcp_fb_nack.pid = -1;
+    stream->rx_last_seq = -1;
+    stream->is_idle = PJ_TRUE;
+
+#if defined(PJMEDIA_STREAM_ENABLE_KA) && PJMEDIA_STREAM_ENABLE_KA!=0
+    c_strm->use_ka = info->use_ka;
+    c_strm->ka_interval = info->ka_cfg.ka_interval;
+    c_strm->start_ka_count = info->ka_cfg.start_count;
+    c_strm->start_ka_interval = info->ka_cfg.start_interval;
+#endif
+
+    c_strm->cname = info->cname;
+    if (c_strm->cname.slen == 0) {
+        /* Build random RTCP CNAME. CNAME has user@host format */
+        c_strm->cname.ptr = p = (char*) pj_pool_alloc(pool, 20);
+        pj_create_random_string(p, 5);
+        p += 5;
+        *p++ = '@'; *p++ = 'p'; *p++ = 'j';
+        pj_create_random_string(p, 6);
+        p += 6;
+        *p++ = '.'; *p++ = 'o'; *p++ = 'r'; *p++ = 'g';
+        c_strm->cname.slen = p - c_strm->cname.ptr;
+    }
+
+    /* Init buffering time and create clock. */
+    stream->buf_time = info->buffer_time;
+    if (stream->buf_time == 0)
+        stream->buf_time = BUFFERING_TIME;
+    cparam.usec_interval = stream->buf_time * 1000;
+    cparam.clock_rate = 1000;
+    status = pjmedia_clock_create2(pool, &cparam,
+                                   PJMEDIA_CLOCK_NO_HIGHEST_PRIO,
+                                   clock_cb, stream, &stream->clock);
```

Comment:
Looking at the clock interval & jitter buffer params (frame size & max cnt), it does not seem optimized for 'medium-sized' image/video/binary files. Anyway, I believe sending binary is not part of the current objectives.

Reply:
Yes, for custom usage (such as binary), users will need to adjust some things. Because the primary usage here is of course text as per the RFC, and the RFC anticipates usage of about 30 cps only (character per second) and recommends a rather long buffering time of 300 ms.

Reply:
Okay, so it looks like transfering binary file needs to be done via other means.